File: Problems/No.141_Linked_List_Cycle/linked_list_cycle.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution(object):
    def hasCycle(self, head):
        """
        :type head: ListNode
        :rtype: bool
        """
        
        # parameters
        fast, slow = head, head
        
        # start compute, using Floyd's Cycle-Finding Algorithm
        while (fast!=None and fast.next!=None):
            fast = fast.next.next
            slow = slow.next
            if (fast == slow):
                return True
        
        return False
        

File: Problems/No.141_Linked_List_Cycle/linked_list_cycle.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    bool hasCycle(ListNode *head) {
        
        // parameters
        ListNode *fast = head;
        ListNode *slow = head;
        
        // start compute, using Floyd's Cycle-Finding Algorithm
        while (fast!=NULL and fast->next!=NULL) {
            
            fast = fast->next->next;
            slow = slow->next;   
            
            if (fast == slow) return true;
            
        }
        
        return false;
        
    }
};
